include/secp256k1_schnorrsig.h

```diff
@@ -0,0 +1,125 @@
+ * BIP-schnorr
+ * (https://github.com/sipa/bips/blob/bip-schnorr/bip-schnorr.mediawiki).
+ */
+
+/** Opaque data structure that holds a parsed Schnorr signature.
+ *
+ *  The exact representation of data inside is implementation defined and not
+ *  guaranteed to be portable between different platforms or versions. It is
+ *  however guaranteed to be 64 bytes in size, and can be safely copied/moved.
+ *  If you need to convert to a format suitable for storage, transmission, or
+ *  comparison, use the `secp256k1_schnorrsig_serialize` and
+ *  `secp256k1_schnorrsig_parse` functions.
+ */
+typedef struct {
+    unsigned char data[64];
+} secp256k1_schnorrsig;
+
+/** Serialize a Schnorr signature.
+ *
+ *  Returns: 1
+ *  Args:    ctx: a secp256k1 context object
+ *  Out:   out64: pointer to a 64-byte array to store the serialized signature
+ *  In:      sig: pointer to the signature
+ *
+ *  See secp256k1_schnorrsig_parse for details about the encoding.
+ */
+SECP256K1_API int secp256k1_schnorrsig_serialize(
+    const secp256k1_context* ctx,
+    unsigned char *out64,
+    const secp256k1_schnorrsig* sig
+) SECP256K1_ARG_NONNULL(1) SECP256K1_ARG_NONNULL(2) SECP256K1_ARG_NONNULL(3);
+
+/** Parse a Schnorr signature.
+ *
+ *  Returns: 1 when the signature could be parsed, 0 otherwise.
+ *  Args:    ctx: a secp256k1 context object
+ *  Out:     sig: pointer to a signature object
+ *  In:     in64: pointer to the 64-byte signature to be parsed
+ *
+ * The signature is serialized in the form R||s, where R is a 32-byte public
+ * key (X coordinate only; the Y coordinate is considered to be the unique
+ * Y coordinate satisfying the curve equation that is square)
+ * and s is a 32-byte big-endian scalar.
+ *
+ * After the call, sig will always be initialized. If parsing failed or the
+ * encoded numbers are out of range, signature validation with it is
+ * guaranteed to fail for every message and public key.
+ */
+SECP256K1_API int secp256k1_schnorrsig_parse(
+    const secp256k1_context* ctx,
+    secp256k1_schnorrsig* sig,
+    const unsigned char *in64
+) SECP256K1_ARG_NONNULL(1) SECP256K1_ARG_NONNULL(2) SECP256K1_ARG_NONNULL(3);
+
+/** Create a Schnorr signature.
+ *
+ * Returns 1 on success, 0 on failure.
+ *  Args:    ctx: pointer to a context object, initialized for signing (cannot be NULL)
+ *  Out:     sig: pointer to the returned signature (cannot be NULL)
+ *  In:    msg32: the 32-byte message being signed (cannot be NULL)
+ *        seckey: pointer to a 32-byte secret key (cannot be NULL)
+ *       noncefp: pointer to a nonce generation function. If NULL, secp256k1_nonce_function_bipschnorr is used
+ *         ndata: pointer to arbitrary data used by the nonce generation function (can be NULL)
+ */
+SECP256K1_API int secp256k1_schnorrsig_sign(
+    const secp256k1_context* ctx,
+    secp256k1_schnorrsig *sig,
+    const unsigned char *msg32,
+    const unsigned char *seckey,
+    secp256k1_nonce_function noncefp,
+    void *ndata
```

Comment:
uh. shouldn't this present an interface that takes the pubkey as an argument, rather than halving the signing speed for data the caller already probably has (and if they don't have it, they can generate it at the same speed with one extra line of code).  Or perhaps accept a null pointer for the case where they don't have it.

Reply:
I was thinking about this too. A complication with x-only public keys is that to take advantage of this, you'd also need to pass in whether the Y coordinate of privkey*G is a quadratic residue or not (as that implies negating the private key before signing).

Reply:
To solve this, we'll probably want a data structure that stores both the x-only pubkey and a boolean that indicates its negation-needed flag (which is equal to whether the privkey*G has a non-square Y coordinate).

That structure starts to look surprisingly much like the old public key type, except it's square/nonsquare Y rather than even/odd.

Reply:
Yes it should be possible to use an interface that's twice as fast if you know the pubkey.

Note that in that case we'll need to add the pubkey as extradata to the deterministic nonce function, otherwise calling sign with the wrong public key may leak the secret key through an (invalid) signature. This makes the signing function with pubkey argument not pass the signing test vectors.

> That structure starts to look surprisingly much like the old public key type

This is what I did originally for `xonly_pubkey_create`, `xonly_tweak_add` etc. But people found that confusing, and so I changed the functions to output `xonly_pubkey` and `is_negated` (but apparently not for the `xonly_pubkey_create` function). `is_negated` is crucial in avoiding similar recomputation in MuSig and the likes, but apparently I missed the normal Schnorr signing use case.

Reply:
Any background on why it was confusing?  A single object (which you'd then have a seperate serializer for) seems less confusing to me.

Reply:
The confusing part was probably my explanation for why this is useful because the interface wasn't as clear at the time. However, I'd be happy to change it back as the `is_negated` flag is a recurring nuisance.

Reply:
I hate to bring this up at this point, but this discussion makes me wonder if we should consider the changing the x-only pubkey tiebreaker from square-y to even-y (but leave the R one at square-y).

First of all, using square-y strictly adds a Jacobi symbol computation to either the keygen or signing process, for no gain except consistency with R. Using square-y for R itself does not have this problem, as it avoids a conversion to affine coordinates, but we always need P in affine coordinates anyway.

However, if we're going to need an application-exposed full pubkey type, using square-y as a tie breaker adds another downside, namely not being able to reuse the existing normal public keys. This matters, because compatibility with existing key generation is a goal, and I believe that interaction between the two may be annoying in some cases.

Think about the use case of a simple single-user Taproot output, and spending it via PSBT. PSBT already has records that say "this (full) pubkey was derived this way" (from which master key, and which key derivation path), which we'll want to reuse in a Taproot world. For key path spending, we'll need a new PSBT record that states "this (xonly) public key is equal to (reference to full pubkey) with tweak X". With square-y keys, if we want to avoid forcing the Jacobi symbol computation onto PSBT signers, it also needs flags to indicate whether the private key is to be negated before the tweak and after the tweak. With even-y keys, a generic "this full public key is this full public key + this tweak" that isn't even Taproot specific suffices.

Reply:
I think user in general should be discouraged from reusing keys in different applications, beyond the general precautions that get given for all key reuse in crypto (surprising bad things can happen)-- there is a very specific easily reached route where using keys across different uses will break things:   Use of schnorr+ecdsa on the same message with the same nonce leaks the key.  I'm confident some implementations are going to screw that up, and reuse their existing rfc6979...

Being compatible with existing generation routines doesn't mean reusing the keys in other applications.

If you take the key reuse off the table I don't think it matters particularly much.  We could, if we wanted, define a new kind of 'compressed pubkey' that instead of 2/3 uses different bits to signal square/non-square.

I agree that it's weak one way vs another.

Reply:
thinking about it, one way would be to use the existing pubkey objects. Add an internal flag for the squareness.  And have serialize flags that know how to serialize either as xonly OR using a non-standard value in the first byte to signal the squareness?

Reply:
Well. I agree, there should be a way to sign without the need to recompute a public key that the user knows already. But this here worries me:

> Note that in that case we'll need to add the pubkey as extradata to the deterministic nonce function, otherwise calling sign with the wrong public key may leak the secret key through an (invalid) signature. This makes the signing function with pubkey argument not pass the signing test vectors.

Yes, and annoyingly we need to add the extended public key here. Otherwise you can trick people into producing two signatures, one with and one without negated signing key, which leaks your nonce by `s1 + s2 = (r + cx) + (r - cx) = 2r`. Note that this is a different attack than the ordinary nonce reuse attack: here, you reuse the nonce `r` but you actually sign for the same hash `c = H(pk||R||m)`. So this is pretty subtle and I bet people will get this wrong. :/

This brings me to a question of provable security: if in practice there is a signing algorithm that takes a public key, then you should also have an equivalent signing oracle in your security model on paper. I don't think it's hard but unfortunately, noone has looked into this.  

It was a deliberate choice to make schnorr pubkeys different from ecdsa pubkeys to discourage reuse across different scheme. And now they are so different that they are even recognizable by their size, which I think is a good thing.  

Let's call this "extended pubkeys". I prefer "extended" over "full" because full suggests that for non-full keys, an essential part is missing. But "extended" it's not great either. Maybe there is a better term.

I would love to propose that extended pubkeys should have a serialization which is different from 2/3. But if we want to be compatible to BIP32, that's just not possible. Or do people think it's a good idea to be compatible to a variant of BIP32 only? That seems somewhat weird to me but one point is that anyway you want different derivation paths for schnorr vs ecdsa...

I was never too convinced about the advantage of "consistency of the tie-breaker". If it's faster, why not use odd/even.

Reply:
Not a big deal, but using the oddness tiebreaker requires adding another line in the BIP's verification algo that negates `y` if odd. We don't need that for the squareness tie breaker because the square root automatically computes the square `y`. This was the original rationale for using squareness (not thinking about signing) in addition to the consistency.

Reply:
Indeed, squareness is faster to decode but a negation is about as close to free as you can get while still doing something.

Reply:
or you make a super-extended public key which also includes a tagged cryptographic hash of the private key. :-/

Reply:
Splitting discussions out. This one about the performance impact on signing, without precomputed public key data.

* With squareness tie-breaker, signing needs to compute (privkey*G), compute its affine X coordinate (=computing inv(Z), squaring it, and multiplying with X), compute the squaredness of its Y coordinate (=computing Y*Z, and jacobi of it), normalize X, and then conditionally negate the privkey.

* With oddness tie-breaker, signing needs to compute (privkey*G), compute its affine X coordinate (=computing inv(Z), squaring it, and multiplying with X), compute the affine Y coordinate (=computing 1/Z^3, and multiplying with Y), normalize X, normalize Y, and then conditionally negate the privkey.

The current PR does one unnecessary multiplication, as it's unnecessarily computing the affine Y coordinate. After fixing that, the speed gain from going from square to even tiebreaker is trading 1 field multiplication and normalization for 1 jacobi computation - which is dominated by the jacobi symbol.

I don't know if we care about this ~5us speedup at signing time (because when you do care, perhaps you want signing with precomputed public key data, which avoids this cost in both cases), but there seems to be basically no gain apart from consistency with R (which is arguably very different, as handling of R is generally limited in scope to the signing code, while public keys are obviously exposed to applications (and things like the taproot tweak break the perfect blackboxy ness of them).

Reply:
Split out discussion. This one is about avoiding the EC multiplication by using precomputed pubkey data.

It seems a question is whether we need to protect against the situation where the public key (or data derived from it), passed to the signing function, is actually computed by the signer itself, or untrusted.

In the case of signing devices with a fixed master BIP32 key, the public key is likely going be computed by the device itself anyway at signing time, so it's wasteful to have the signing algorithm redo it.

On the other hand, @real-or-random and @jonasnick correctly point out that if the public key data provided to the signing algorithm is untrusted, this public key data must also be an input to the derivation function. I don't think we have a proof that even with this modified derivation function, the resulting signature scheme is secure.

So I think that's an unfortunate situation to be in. We shouldn't standardize an (alternative) signing algorithm that takes untrusted pubkey data as input (because we have no proof it is secure), but simultaneously, having an API that takes in trusted pubkey data is scary (because if the API is abused with untrusted pubkey data, we have a known attack).

FWIW, the ed25519-donna implementation seems to happily take the public key as input to the signer. I wonder if the concern of an untrusted public key has been analysed in that context.

Reply:
@sipa I took the freedom to "move" your two summary posts to https://github.com/sipa/bips/issues/190 and https://github.com/sipa/bips/issues/191 because GitHub's "comment on code" feature is not good for extended discussions, and these issues are actually issues, and they relate to the BIP itself, not only the implementation. 

Please continue there.

Reply:
May I come back to the original discussion of should `is_negated` be part of the pubkey part and serialized out or not?
I mean the only need for `is_negated` is to verify that you tweak correctly, can/should this be part of the struct? if so it will mean that `tweak_test(der(ser(add(key, tweak), key, tweak) != tweak_test(add(key,tweak), key, tweak)`

Reply:
@elichai Not exactly sure what your point is but we have a struct that has the `is_negated` part and is serializable - the plain old `secp256k1_pubkey` struct. Sounds like you're arguing for API design option 2 (https://github.com/bitcoin-core/secp256k1/pull/558#issuecomment-582152117). I don't think compressed serialization is not enough, because taproot users need a single `is_negated` bit to stuff in the control block and I don't think we want them to extract it themselves from the `0x02/0x03` tags.

Reply:
yeah I guess we must output the oddness byte someone, so we might as well have it as a separate thing instead of compressed because for most sign/verify you don't need it

Reply:
I think this is resolved now the function takes a `secp256k1_keypair` as input.